lib/bolt/logger.rb

```diff
@@ -123,18 +130,54 @@ def self.reset_logging
       Logging.reset
     end
 
-    def self.warn_once(type, msg)
+    # Issues a warning unless the user has suppressed it.
+    def self.warn(id:, msg:, src: self)
+      return if @disable_warnings.include?(id)
+
       @mutex.synchronize do
-        @logger ||= Bolt::Logger.logger(self)
-        if @warnings.add?(type)
-          @logger.warn(msg)
+        Bolt::Logger.logger(src).warn("#{msg} [ID: #{id}]")
+      end
+    end
+
+    # Issues a warning unless the user has suppressed it,
+    # or a warning with the same ID has already been issued.
+    def self.warn_once(id:, msg:, src: self)
+      return if @disable_warnings.include?(id)
+
+      @mutex.synchronize do
+        if @warnings.add?(id)
+          Bolt::Logger.logger(src).warn("#{msg} [ID: #{id}]")
         end
       end
     end
 
-    def self.deprecation_warning(type, msg)
-      @analytics&.event('Warn', 'deprecation', label: type)
-      warn_once(type, msg)
+    # Issues a deprecation warning unless the user has suppressed
+    # it. Submits analytics for the deprecation regardless of
+    # whether the warning is suppressed.
+    def self.deprecation_warning(id:, msg:, src: self)
+      @analytics&.event('Warn', 'deprecation', label: id)
+      warn(id: id, msg: msg, src: src)
+    end
+
+    # Queues a log message to be issued at a later time. This is used when
+    # parsing the command and loading the project/config, as the logger
+    # is not configured until after that happens.
+    def self.queue_message(type:, msg:, id: nil, src: nil)
+      @message_queue << { type: type, msg: msg, id: id, src: src || self }
+    end
+
+    # Issues log messages in the message queue.
+    def self.issue_queued_messages
+      @message_queue.each do |message|
+        case message[:type]
+        when :deprecation
+          deprecation_warning(id: message[:id], msg: message[:msg], src: message[:src])
+        when :warn
+          warn(id: message[:id], msg: message[:msg], src: message[:src])
+        else
+          logger(message[:src]).send(message[:type], message[:msg])
+        end
+      end
```

Comment:
While not strictly necessary to enable this change, these methods help to simplify handling messages logged before the logger is configured. I don't believe this should cause any issues / race conditions, as the `Logging` library handles creating `Logging` instances.